src/Std/Internal/Http/Data/Request.lean

```diff
@@ -0,0 +1,226 @@
+
+public section
+
+namespace Std
+namespace Http
+
+set_option linter.all true
+
+open Lean
+
+/--
+The main parts of a request containing the HTTP method, version, URI, and headers.
+-/
+structure Request.Head where
+  /--
+  The HTTP method (GET, POST, PUT, DELETE, etc.) for the request
+  -/
+  method : Method := .get
+
+  /--
+  The HTTP protocol version (HTTP/1.0, HTTP/1.1, HTTP/2.0, etc.)
+  -/
+  version : Version := .v11
+
+  /--
+  The request target/URI indicating the resource being requested
+  -/
+  uri : RequestTarget := .asteriskForm
+
+  /--
+  Collection of HTTP headers for the request (Content-Type, Authorization, etc.)
+  -/
+  headers : Headers := .empty
+deriving Inhabited, Repr
+
+/--
+HTTP request structure parameterized by body type
+-/
+structure Request (t : Type) where
+  /--
+  The request headers and metadata
+  -/
+  head : Request.Head
+
+  /--
+  The request body content of type t
+  -/
+  body : t
+deriving Inhabited
+
+/--
+Builds a HTTP Request
+-/
+structure Request.Builder where
+
+  /--
+  The head of the request
+  -/
+  head : Head := {}
+
+namespace Request
+
+instance : ToString Head where
+  toString req :=
+    toString req.method ++ " " ++
+    toString req.uri ++ " " ++
+    toString req.version ++
+    "\r\n" ++
+    toString req.headers ++ "\r\n\r\n"
+
+/--
+Determines if a request method typically doesn't allow a request body
+-/
+@[inline]
+def isInformational (request : Head) : Prop :=
+  ¬request.method.allowsRequestBody
+
+/--
+Creates a new HTTP Request builder with default head (method: GET, version: HTTP/1.1, asterisk URI, empty headers)
+-/
+def new : Builder := { }
+
+namespace Builder
+
+/--
+Creates a new HTTP Request builder with default head (method: GET, version: HTTP/1.1, asterisk URI, empty headers)
+-/
+def empty : Builder := { }
+
+/--
+Sets the HTTP method for the request being built
+-/
+def method (builder : Builder) (method : Method) : Builder :=
+  { builder with head := { builder.head with method := method } }
+
+/--
+Sets the HTTP version for the request being built
+-/
+def version (builder : Builder) (version : Version) : Builder :=
+  { builder with head := { builder.head with version := version } }
+
+/--
+Sets the request target/URI for the request being built
+-/
+def uri (builder : Builder) (uri : RequestTarget) : Builder :=
+  { builder with head := { builder.head with uri := uri } }
+
+/--
+Sets the request target/URI for the request being built
+-/
+def uri! (builder : Builder) (uri : String) : Builder :=
+  { builder with head := { builder.head with uri := RequestTarget.parse! uri } }
+
+/--
+Adds a single header to the request being built
+-/
+def header (builder : Builder) (key : String) (value : String) : Builder :=
+  { builder with head := { builder.head with headers := builder.head.headers.insert key value } }
+
+/--
+Builds and returns the final HTTP Request with the specified body
+-/
+def body (builder : Builder) (body : t) : Request t :=
```

Comment:
Should this function force the user to set the `Content-Type` header?

Reply:
I think it's something the HTTP client can infer and add when someone tries to send the request, to prevent it from being malformed.

Reply:
Fixed by adding `ContentType` type class.